src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -397,27 +364,208 @@ CollectionExpressionTypeKind.Span or
                 span.Type);
         }
 
-        private BoundExpression MakeCollectionExpressionSpreadElement(BoundCollectionExpressionSpreadElement initializer)
+        /// <summary>
+        /// Create and populate an array from a collection expression where the
+        /// collection has a known length, although possibly including spreads.
+        /// </summary>
+        private BoundExpression CreateAndPopulateArray(BoundCollectionExpression node, ArrayTypeSymbol arrayType)
         {
-            var enumeratorInfo = initializer.EnumeratorInfoOpt;
-            var addElementPlaceholder = initializer.AddElementPlaceholder;
+            var syntax = node.Syntax;
+            var elements = node.Elements;
 
-            Debug.Assert(enumeratorInfo is { });
-            Debug.Assert(addElementPlaceholder is { });
-            Debug.Assert(addElementPlaceholder.Type is { });
+            if (node.GetKnownLength(out bool hasSpreadElements) is not int knownLength)
+            {
+                // Should have been handled by the caller.
+                throw ExceptionUtilities.UnexpectedValue(node);
+            }
 
-            var syntax = (CSharpSyntaxNode)initializer.Syntax;
-            var iterationVariable = _factory.SynthesizedLocal(addElementPlaceholder.Type, syntax);
-            var convertedExpression = (BoundConversion)initializer.Expression;
+            if (knownLength == 0)
+            {
+                return CreateEmptyArray(syntax, arrayType);
+            }
 
-            AddPlaceholderReplacement(addElementPlaceholder, _factory.Local(iterationVariable));
+            if (!hasSpreadElements)
+            {
+                var initialization = new BoundArrayInitialization(
+                        syntax,
+                        isInferred: false,
+                        elements.SelectAsArray(e => VisitExpression(e)));
```

Comment:
Consider making this static, and passing through `this` as a parameter, so the lambda can be cached.